src/main/java/fr/gantoin/technical_basic_test/controller/ProductController.java

```diff
@@ -40,19 +39,29 @@ private Map<Integer, Product> initProducts() {
 
     @GetMapping("/products")
     public List<Product> getProducts() {
-        // TODO: Convert the map to a list and return it
-        return existingProducts;
+        return existingProducts.values().stream().toList();
     }
 
     @GetMapping("/products/{id}")
-    public Product getProductById(@PathVariable int id) {
-        //TODO: Implement the method to return the product with the given id
-        return null;
+    public ResponseEntity<Product> getProductById(@PathVariable int id) {
+        Product product = existingProducts.get(id);
+
+        if (product == null) {
+            return ResponseEntity.notFound().eTag("No such product id").build();
+        }
+
+        return ResponseEntity.ok(product);
     }
 
     @PostMapping("/products")
-    public Product addProduct(@RequestBody Product product) {
-        // TODO: Implement the method to add a new product and return it
-        return null;
+    public ResponseEntity<Product> addProduct(@RequestBody Product product) {
+        int lastId = existingProducts.keySet().stream().max(Integer::compareTo).orElse(0);
+
+        if (product.id() <= lastId) {
+            return ResponseEntity.notFound().eTag("Wrong id").build();
+        }
```

Comment:
Could be also something like:
```suggestion
        if (existingProducts.containsKey(product.id())) {
            return ResponseEntity.badRequest().eTag("Product already exists").build();
        }
```

If we do not expect incremental ids sequence.